llvm/lib/CheerpWriter/CheerpWasmWriter.cpp

```diff
@@ -3559,6 +3607,85 @@ bool CheerpWasmWriter::compileInlineInstruction(WasmBuffer& code, const Instruct
 			}
 			break;
 		}
+		case Instruction::AtomicRMW:
+		{
+			const AtomicRMWInst& ai = cast<AtomicRMWInst>(I);
+			const Type* opType = ai.getOperand(1)->getType();
+			assert(opType->isIntegerTy());
+			const Value* ptrOp = ai.getPointerOperand();
+			const Value* valOp = ai.getValOperand();
+			uint32_t offset = compileLoadStorePointer(code, ptrOp);
+			compileOperand(code, valOp);
+			switch (ai.getOperation())
+			{
+#define ATOMICBINOP(Ty, name) \
+				case AtomicRMWInst::Ty: \
+				{ \
+					if (opType->isIntegerTy(64)) \
+						encodeInst(WasmThreadsU32U32Opcode::I64_ATOMIC_RMW_##name, 0x3, offset, code); \
+					else if (opType->isIntegerTy(32)) \
+						encodeInst(WasmThreadsU32U32Opcode::I32_ATOMIC_RMW_##name, 0x2, offset, code); \
+					else if (opType->isIntegerTy(16)) \
+						encodeInst(WasmThreadsU32U32Opcode::I32_ATOMIC_RMW16_##name##_U, 0x1, offset, code); \
+					else if (opType->getPrimitiveSizeInBits() <= 8) \
+						encodeInst(WasmThreadsU32U32Opcode::I32_ATOMIC_RMW8_##name##_U, 0x0, offset, code); \
+					else \
+						llvm::report_fatal_error("Unknown bitwidth for AtomicRMW inst"); \
+					break; \
+				}
+				ATOMICBINOP( Add,  ADD)
+				ATOMICBINOP( Sub,  SUB)
+				ATOMICBINOP( And,  AND)
+				ATOMICBINOP(  Or,   OR)
+				ATOMICBINOP( Xor,  XOR)
+				ATOMICBINOP(Xchg, XCHG)
+#undef ATOMICBINOP
+				default:
+				{
+					llvm::report_fatal_error("Atomic opcode not supported");
+				}
+			}
+			break;
+		}
+		case Instruction::AtomicCmpXchg:
+		{
+			const AtomicCmpXchgInst& ai = cast<AtomicCmpXchgInst>(I);
+			const Value* ptrOp = ai.getPointerOperand();
+			const Value* compareOp = ai.getCompareOperand();
+			const Value* newValOp = ai.getNewValOperand();
+			const Type* t = compareOp->getType();
+			uint32_t offset = compileLoadStorePointer(code, ptrOp);
+			compileOperand(code, compareOp);
```

Comment:
add a comment here and below that says that we are rendering compareOp twice but it's ok because it will not be inlined when usedby cmpxchg

Reply:
Done